RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc

```diff
@@ -0,0 +1,85 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/Common/interface/OwnVector.h"
+#include "RecoTracker/TkTrackingRegions/interface/TrackingRegion.h"
+#include "RecoTracker/TkTrackingRegions/interface/RectangularEtaPhiTrackingRegion.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+
+#include <vector>
+#include <memory>
+
+namespace {
+  class TrackSelectorByRegion final : public edm::global::EDProducer<> {
+  public:
+    explicit TrackSelectorByRegion(const edm::ParameterSet& conf)
+        : tracksToken_(consumes<reco::TrackCollection>(conf.getParameter<edm::InputTag>("tracks"))) {
+      inputTrkRegionToken_ = consumes<edm::OwnVector<TrackingRegion>>(conf.getParameter<edm::InputTag>("inputROI"));
+      produce_collection = conf.getParameter<bool>("produceTrackCollection");
+      if(produce_collection) produces<reco::TrackCollection>();
+      produces<std::vector<bool>>();
+    }
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+      desc.add<edm::InputTag>("tracks", edm::InputTag("hltPixelTracks"));
+      desc.add<edm::InputTag>("inputROI", edm::InputTag(""));
+      desc.add<bool>("produceTrackCollection", true);
+      descriptions.add("trackSelectorByRegion", desc);
+    }
+
+  private:
+    using MaskCollection = std::vector<bool>;
+
+    void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup&) const override {
+      // products
+      auto mask = std::make_unique<MaskCollection>();                  // mask w/ the same size of the input collection
+      auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
+
+      auto regionsHandle = iEvent.getHandle(inputTrkRegionToken_);
+
+      auto tracksHandle = iEvent.getHandle(tracksToken_);
+
+      if (tracksHandle.isValid()) {
+        const auto tracks = *tracksHandle;
+        mask->assign(tracks.size(), false);
+        if (regionsHandle.isValid()) {
+          const auto& regions = *regionsHandle;
+
+          for (const auto& tmp : regions)
+            if (const auto* roi = dynamic_cast<const TrackingRegion*>(&tmp)) {
+              auto amask = roi->checkTracks(tracks);
+              for (size_t it = 0; it < amask.size(); it++) {
+                mask->at(it) = mask->at(it) or amask.at(it);
+              }
+            }
+        }
+
+        assert(mask->size() == tracks.size());
+
+        for (size_t it = 0; it < mask->size(); it++)
+          if (mask->at(it))
```

Comment:
I would move this in the above loop, in order to avoid to have 2 loops

Reply:
Actually I think it's better to have a separate loop: if you do in the loop above, you need to check that a track is not inserted multiple times if it matches multiple regions.